play-json/shared/src/main/scala/play/api/libs/json/JsLookupResult.scala

```diff
@@ -0,0 +1,81 @@
+/*
+ * Copyright (C) 2009-2017 Lightbend Inc. <https://www.lightbend.com>
+ */
+
+package play.api.libs.json
+
+sealed trait JsLookupResult extends Any with JsReadable {
+  /**
+   * Tries to convert the node into a JsValue
+   */
+  def toOption: Option[JsValue]
+
+  def toEither: Either[JsonValidationError, JsValue]
+
+  def get: JsValue = toOption.get
+
+  def getOrElse(v: => JsValue): JsValue = toOption.getOrElse(v)
+  def isEmpty: Boolean
+  def isDefined: Boolean = !isEmpty
+
+  /**
+   * If this result is defined return `this`. Otherwise return `alternative`.
+   */
+  def orElse(alternative: => JsLookupResult): JsLookupResult =
+    if (isDefined) this else alternative
+
+  /**
+   * If this result contains `JsNull` or is undefined, returns `JsSuccess(None)`.
+   * Otherwise returns the result of validating as an `A` and wrapping the result in a `Some`.
+   */
+  def validateOpt[A](implicit rds: Reads[A]): JsResult[Option[A]]
+
+}
+object JsLookupResult {
+  import scala.language.implicitConversions
+  implicit def jsLookupResultToJsLookup(value: JsLookupResult): JsLookup = JsLookup(value)
+}
+
+/**
+ * Wrapper for JsValue to represent an existing Json value.
+ */
+case class JsDefined(value: JsValue) extends AnyVal with JsLookupResult {
+
+  def toOption: Option[JsValue] = Some(value)
+
+  def toEither: Either[JsonValidationError, JsValue] = Right(value)
+
+  def isEmpty: Boolean = false
```

Comment:
Would prefer `isEmpty` `toEither` `toOption` as `val`

Reply:
Doesn't that just mean we're storing the value in two places now? I don't see the benefit.

Reply:
For me it just means that the different representations of the pure value are only instantiated once (e.g. no different `Some` for each call to `toOption`). Could also be `lazy val`, but I see few benefit there.